File: flutter/lib/views/loginViewModel.dart
import 'package:flutter/material.dart';
import '../data/database.dart';
import '../data/user/userDao.dart';
import '../data/user/userRepository.dart';
import '../data/user/userService.dart';
import '../domain/user/usecases/getAllUsersUseCase.dart';
import '../domain/user/usecases/loginUseCase.dart';
import '../domain/user/login.dart';
import '../domain/user/user.dart';
import 'bindingTextField.dart';

class LoginViewModel extends ChangeNotifier {
  Binding<String> email = Binding.initialValue("");
  Binding<String> password = Binding.initialValue("");
  String error;
  Function goToNextScreen;

  UserRepositoryImpl userRepository;

  LoginViewModel() {
    userRepository = UserRepositoryImpl(UserDao(AppDatabase()), UserService());
  }

  void loadData() async {
    try {
      List<User> users = await GetAllUsersUseCase(userRepository).run();
      if (users != null && users.length > 0) {
        setLoggedUserAndProccedToNextScreen(users[0]);
      }
    } catch (error) {
      this.error = error.message;
    }
  }

  void setLoggedUserAndProccedToNextScreen(User user) {
    User.currentUser = user;
    goToNextScreen();
  }

  void doLogin() async {
    Login login = Login(username: this.email.value.trim(), password: this.password.value.trim());
    try {
      User loggedUser = await LoginUseCase(userRepository).run(login);
      setLoggedUserAndProccedToNextScreen(loggedUser);
    } catch (error) {
      this.error = error.message;
      notifyListeners();
    }
  }
}

File: web/src/views/Login/LoginViewModel.ts
import UserRepository from '../../domain/user/UserRepository';
import UserService from '../../data/UserService';
import UserRepositoryImpl from '../../data/UserRepositoryImpl';
import Login from '../../domain/user/Login';
import LoginUseCase from '../../domain/user/usecases/LoginUseCase';
import ViewModel from '../ViewModel';

export type State = {
    name: string;
    password: string;
    error: string | null;
    loading: boolean;
}

export class LoginViewModel extends ViewModel {
    onLogin: Function | null = null;
    userRepository: UserRepository;

    constructor() {
        super();
        this.userRepository = new UserRepositoryImpl(new UserService());
        this.state = {
            name: '',
            password: '',
            error: null,
            loading: false,
        }
    }

    async login() {
        const login = new Login(this.state.name, this.state.password);
        try {
            const user = await new LoginUseCase(this.userRepository).run(login);
            console.log(user);
            if (this.onLogin) {
                this.onLogin();
            }
        } catch (e) {
            console.log(e);
            this.setState({
                error: e.message,
            });
        }
    }

    setUserName = (newUsername: string) => {
        this.setState({
            name: newUsername,
        });
    }

    setPassword = (newPassword: string) => {
        this.setState({
            password: newPassword,
        });
    }
}